### ISSUE TITLE
Build fails with Python 3.11

### ISSUE THREAD

**Issue by davide125** [facebook/sapling#144]
This currently fails to build on Fedora Rawhide, which uses Python 3.11

```
/builddir/build/BUILD/sapling-20221115-080554-34470671/eden/scm/build/traitlets-4.3.3-py2.py3-none-any/traitlets/config/loader.py:795: SyntaxWarning: "is" with a literal. Did you mean "=="?
  if len(key) is 1:
/builddir/build/BUILD/sapling-20221115-080554-34470671/eden/scm/build/traitlets-4.3.3-py2.py3-none-any/traitlets/config/loader.py:804: SyntaxWarning: "is" with a literal. Did you mean "=="?
  if len(key) is 1:
edenscm/cext/mpatch.c: In function 'cpygetitem':
edenscm/cext/mpatch.c:66:3: warning: 'PyObject_AsCharBuffer' is deprecated [-Wdeprecated-declarations]
   66 |   if (PyObject_AsCharBuffer(tmp, &buffer, (Py_ssize_t*)&blen))
      |   ^~
In file included from /usr/include/python3.11/Python.h:100,
                 from edenscm/cext/mpatch.c:31:
/usr/include/python3.11/abstract.h:326:17: note: declared here
  326 | PyAPI_FUNC(int) PyObject_AsCharBuffer(PyObject *obj,
      |                 ^~~~~~~~~~~~~~~~~~~~~
edenscm/cext/mpatch.c: In function 'patches':
edenscm/cext/mpatch.c:94:3: warning: 'PyObject_AsCharBuffer' is deprecated [-Wdeprecated-declarations]
   94 |   if (PyObject_AsCharBuffer(text, &in, &inlen))
      |   ^~
/usr/include/python3.11/abstract.h:326:17: note: declared here
  326 | PyAPI_FUNC(int) PyObject_AsCharBuffer(PyObject *obj,
      |                 ^~~~~~~~~~~~~~~~~~~~~
edenscm/cext/dirs.c: In function '_addpath':
edenscm/cext/dirs.c:79:7: warning: 'ob_shash' is deprecated [-Wdeprecated-declarations]
   79 |       ((PyBytesObject*)key)->ob_shash = -1;
      |       ^
In file included from /usr/include/python3.11/bytesobject.h:62,
                 from /usr/include/python3.11/Python.h:50,
                 from edenscm/cext/dirs.c:18:
/usr/include/python3.11/cpython/bytesobject.h:7:35: note: declared here
    7 |     Py_DEPRECATED(3.11) Py_hash_t ob_shash;
      |                                   ^~~~~~~~
edenscm/cext/dirs.c:88:18: error: lvalue required as left operand of assignment
   88 |     Py_SIZE(key) = pos;
      |                  ^
error: command '/usr/lib64/ccache/gcc' failed with exit code 1
make: *** [Makefile:75: oss] Error 1
```

**Comment by davide125** [comment]
Updating the traitlets dependency would probably fix this.

**Comment by davide125** [comment]
Nope, that was a red herring, this failure is due to sapling itself, not traitlets.

**Comment by davide125** [comment]
Here's a better error trace (with make -j1):

```
building 'edenscmnative.mpatch' extension
gcc -Wsign-compare -DDYNAMIC_ANNOTATIONS_ENABLED=1 -DNDEBUG -O2 -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fstack-protector-strong -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -D_GNU_SOURCE -fPIC -fwrapv -O2 -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fstack-protector-strong -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -D_GNU_SOURCE -fPIC -fwrapv -O2 -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fstack-protector-strong -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -D_GNU_SOURCE -fPIC -fwrapv -O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -fPIC -I. -I../.. -I/usr/local/include -I/usr/include/python3.11 -c edenscm/cext/mpatch.c -o build/temp.linux-x86_64-3.11/edenscm/cext/mpatch.o
edenscm/cext/mpatch.c: In function 'cpygetitem':
edenscm/cext/mpatch.c:66:3: warning: 'PyObject_AsCharBuffer' is deprecated [-Wdeprecated-declarations]
   66 |   if (PyObject_AsCharBuffer(tmp, &buffer, (Py_ssize_t*)&blen))
      |   ^~
In file included from /usr/include/python3.11/Python.h:100,
                 from edenscm/cext/mpatch.c:31:
/usr/include/python3.11/abstract.h:326:17: note: declared here
  326 | PyAPI_FUNC(int) PyObject_AsCharBuffer(PyObject *obj,
      |                 ^~~~~~~~~~~~~~~~~~~~~
edenscm/cext/mpatch.c: In function 'patches':
edenscm/cext/mpatch.c:94:3: warning: 'PyObject_AsCharBuffer' is deprecated [-Wdeprecated-declarations]
   94 |   if (PyObject_AsCharBuffer(text, &in, &inlen))
      |   ^~
/usr/include/python3.11/abstract.h:326:17: note: declared here
  326 | PyAPI_FUNC(int) PyObject_AsCharBuffer(PyObject *obj,
      |                 ^~~~~~~~~~~~~~~~~~~~~
gcc -Wsign-compare -DDYNAMIC_ANNOTATIONS_ENABLED=1 -DNDEBUG -O2 -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fstack-protector-strong -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -D_GNU_SOURCE -fPIC -fwrapv -O2 -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fstack-protector-strong -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -D_GNU_SOURCE -fPIC -fwrapv -O2 -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fstack-protector-strong -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -D_GNU_SOURCE -fPIC -fwrapv -O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -fPIC -I. -I../.. -I/usr/local/include -I/usr/include/python3.11 -c edenscm/mpatch.c -o build/temp.linux-x86_64-3.11/edenscm/mpatch.o
gcc -shared -Wl,-z,relro -Wl,--as-needed -Wl,-z,now -Wl,--build-id=sha1 -g -Wl,-z,relro -Wl,--as-needed -Wl,-z,now -Wl,--build-id=sha1 -g -Wl,-z,relro -Wl,--as-needed -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -Wl,--build-id=sha1 -specs=/usr/lib/rpm/redhat/redhat-package-notes -O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection build/temp.linux-x86_64-3.11/edenscm/cext/mpatch.o build/temp.linux-x86_64-3.11/edenscm/mpatch.o -L/usr/lib64 -Lbuild/temp.linux-x86_64-3.11 -lmpatch -lchg -o /builddir/build/BUILD/sapling-20221115-080554-34470671/eden/scm/edenscmnative/mpatch.cpython-311-x86_64-linux-gnu.so
building 'edenscmnative.parsers' extension
gcc -Wsign-compare -DDYNAMIC_ANNOTATIONS_ENABLED=1 -DNDEBUG -O2 -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fstack-protector-strong -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -D_GNU_SOURCE -fPIC -fwrapv -O2 -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fstack-protector-strong -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -D_GNU_SOURCE -fPIC -fwrapv -O2 -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fstack-protector-strong -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -D_GNU_SOURCE -fPIC -fwrapv -O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -fPIC -I. -I../.. -I/usr/local/include -I/usr/include/python3.11 -c edenscm/cext/charencode.c -o build/temp.linux-x86_64-3.11/edenscm/cext/charencode.o
gcc -Wsign-compare -DDYNAMIC_ANNOTATIONS_ENABLED=1 -DNDEBUG -O2 -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fstack-protector-strong -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -D_GNU_SOURCE -fPIC -fwrapv -O2 -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fstack-protector-strong -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -D_GNU_SOURCE -fPIC -fwrapv -O2 -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fstack-protector-strong -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -D_GNU_SOURCE -fPIC -fwrapv -O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -fPIC -I. -I../.. -I/usr/local/include -I/usr/include/python3.11 -c edenscm/cext/dirs.c -o build/temp.linux-x86_64-3.11/edenscm/cext/dirs.o
edenscm/cext/dirs.c: In function '_addpath':
edenscm/cext/dirs.c:79:7: warning: 'ob_shash' is deprecated [-Wdeprecated-declarations]
   79 |       ((PyBytesObject*)key)->ob_shash = -1;
      |       ^
In file included from /usr/include/python3.11/bytesobject.h:62,
                 from /usr/include/python3.11/Python.h:50,
                 from edenscm/cext/dirs.c:18:
/usr/include/python3.11/cpython/bytesobject.h:7:35: note: declared here
    7 |     Py_DEPRECATED(3.11) Py_hash_t ob_shash;
      |                                   ^~~~~~~~
edenscm/cext/dirs.c:88:18: error: lvalue required as left operand of assignment
   88 |     Py_SIZE(key) = pos;
      |                  ^
error: command '/usr/lib64/ccache/gcc' failed with exit code 1
make: *** [Makefile:75: oss] Error 1
```

**Comment by davide125** [comment]
I can confirm this builds fine on Fedora 36 (which uses Python 3.10), so it's specifically Python 3.11 incompatibility.

**Comment by safarmer** [comment]
Looks like the Py_SIZE API has changed.

From Python docs:
> [Py_ssize_t](https://docs.python.org/3/c-api/intro.html#c.Py_ssize_t) Py_SIZE([PyVarObject](https://docs.python.org/3/c-api/structures.html#c.PyVarObject) *o)
> Get the size of the Python object o.
> 
> Use the [Py_SET_SIZE()](https://docs.python.org/3/c-api/structures.html#c.Py_SET_SIZE) function to set an object size.
> 
> Changed in version 3.11: [Py_SIZE()](https://docs.python.org/3/c-api/structures.html#c.Py_SIZE) is changed to an inline static function. The parameter type is no longer const [PyVarObject](https://docs.python.org/3/c-api/structures.html#c.PyVarObject)*.

I think this is the least of the problems with the internal changes to the Python frame object handling: https://github.com/python/cpython/issues/88756

TL;DR: the `traceprof` related code needs changes to how `PyFrameObject` is used (no longer a complete struct/type in the currently used headers). I am not a Python expert so I am not sure how to approach this.

**Comment by quark-zju** [comment]
Not resolved. There are other users of `Py_SIZE`.

**Comment by leahneukirchen** [comment]
Fixing Py_SIZE, it goes on with different issues:

```
./edenscm/ext/extlib/traceprofimpl.cpp: In member function 'void FrameInfo::assign(PyFrameObject*, frameid_t)':
./edenscm/ext/extlib/traceprofimpl.cpp:52:17: error: invalid use of incomplete type 'PyFrameObject' {aka 'struct _frame'}
   52 |     code = frame->f_code;
```

later:

```
edenscmnative/traceprof.cpp: In function 'int __Pyx__GetException(PyThreadState*, PyObject**, PyObject**, PyObject**)':
edenscmnative/traceprof.cpp:3923:30: error: '_PyErr_StackItem' {aka 'struct _err_stackitem'} has no member named 'exc_type'
 3923 |         tmp_type = exc_info->exc_type;
```

**Comment by jtagcat** [comment]
Build still failing with `main`
```
gcc -Wsign-compare -DDYNAMIC_ANNOTATIONS_ENABLED=1 -DNDEBUG -O2 -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fstack-protector-strong -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -D_GNU_SOURCE -fPIC -fwrapv -O2 -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fstack-protector-strong -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -D_GNU_SOURCE -fPIC -fwrapv -O2 -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fstack-protector-strong -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -D_GNU_SOURCE -fPIC -fwrapv -fPIC -I. -I../.. -I/usr/local/include -I/opt/local/include -I/usr/include/python3.11 -c edenscmnative/traceprof.cpp -o build/temp.linux-x86_64-cpython-311/edenscmnative/traceprof.o -std=c++11 -g
edenscmnative/traceprof.cpp: In function 'int __Pyx__GetException(PyThreadState*, PyObject**, PyObject**, PyObject**)':
edenscmnative/traceprof.cpp:3924:30: error: '_PyErr_StackItem' {aka 'struct _err_stackitem'} has no member named 'exc_type'
 3924 |         tmp_type = exc_info->exc_type;
      |                              ^~~~~~~~
edenscmnative/traceprof.cpp:3926:28: error: '_PyErr_StackItem' {aka 'struct _err_stackitem'} has no member named 'exc_traceback'
 3926 |         tmp_tb = exc_info->exc_traceback;
      |                            ^~~~~~~~~~~~~
edenscmnative/traceprof.cpp:3927:19: error: '_PyErr_StackItem' {aka 'struct _err_stackitem'} has no member named 'exc_type'
 3927 |         exc_info->exc_type = local_type;
      |                   ^~~~~~~~
edenscmnative/traceprof.cpp:3929:19: error: '_PyErr_StackItem' {aka 'struct _err_stackitem'} has no member named 'exc_traceback'
 3929 |         exc_info->exc_traceback = local_tb;
      |                   ^~~~~~~~~~~~~
edenscmnative/traceprof.cpp: In function 'void __Pyx__ExceptionSwap(PyThreadState*, PyObject**, PyObject**, PyObject**)':
edenscmnative/traceprof.cpp:3962:26: error: '_PyErr_StackItem' {aka 'struct _err_stackitem'} has no member named 'exc_type'
 3962 |     tmp_type = exc_info->exc_type;
      |                          ^~~~~~~~
edenscmnative/traceprof.cpp:3964:24: error: '_PyErr_StackItem' {aka 'struct _err_stackitem'} has no member named 'exc_traceback'
 3964 |     tmp_tb = exc_info->exc_traceback;
      |                        ^~~~~~~~~~~~~
edenscmnative/traceprof.cpp:3965:15: error: '_PyErr_StackItem' {aka 'struct _err_stackitem'} has no member named 'exc_type'
 3965 |     exc_info->exc_type = *type;
      |               ^~~~~~~~
edenscmnative/traceprof.cpp:3967:15: error: '_PyErr_StackItem' {aka 'struct _err_stackitem'} has no member named 'exc_traceback'
 3967 |     exc_info->exc_traceback = *tb;
      |               ^~~~~~~~~~~~~
edenscmnative/traceprof.cpp: In function '_PyErr_StackItem* __Pyx_PyErr_GetTopmostException(PyThreadState*)':
edenscmnative/traceprof.cpp:3997:23: error: '_PyErr_StackItem' {aka 'struct _err_stackitem'} has no member named 'exc_type'
 3997 |     while ((exc_info->exc_type == NULL || exc_info->exc_type == Py_None) &&
      |                       ^~~~~~~~
edenscmnative/traceprof.cpp:3997:53: error: '_PyErr_StackItem' {aka 'struct _err_stackitem'} has no member named 'exc_type'
 3997 |     while ((exc_info->exc_type == NULL || exc_info->exc_type == Py_None) &&
      |                                                     ^~~~~~~~
edenscmnative/traceprof.cpp: In function 'void __Pyx__ExceptionSave(PyThreadState*, PyObject**, PyObject**, PyObject**)':
edenscmnative/traceprof.cpp:4011:23: error: '_PyErr_StackItem' {aka 'struct _err_stackitem'} has no member named 'exc_type'
 4011 |     *type = exc_info->exc_type;
      |                       ^~~~~~~~
edenscmnative/traceprof.cpp:4013:21: error: '_PyErr_StackItem' {aka 'struct _err_stackitem'} has no member named 'exc_traceback'
 4013 |     *tb = exc_info->exc_traceback;
      |                     ^~~~~~~~~~~~~
edenscmnative/traceprof.cpp: In function 'void __Pyx__ExceptionReset(PyThreadState*, PyObject*, PyObject*, PyObject*)':
edenscmnative/traceprof.cpp:4027:26: error: '_PyErr_StackItem' {aka 'struct _err_stackitem'} has no member named 'exc_type'
 4027 |     tmp_type = exc_info->exc_type;
      |                          ^~~~~~~~
edenscmnative/traceprof.cpp:4029:24: error: '_PyErr_StackItem' {aka 'struct _err_stackitem'} has no member named 'exc_traceback'
 4029 |     tmp_tb = exc_info->exc_traceback;
      |                        ^~~~~~~~~~~~~
edenscmnative/traceprof.cpp:4030:15: error: '_PyErr_StackItem' {aka 'struct _err_stackitem'} has no member named 'exc_type'
 4030 |     exc_info->exc_type = type;
      |               ^~~~~~~~
edenscmnative/traceprof.cpp:4032:15: error: '_PyErr_StackItem' {aka 'struct _err_stackitem'} has no member named 'exc_traceback'
 4032 |     exc_info->exc_traceback = tb;
      |               ^~~~~~~~~~~~~
edenscmnative/traceprof.cpp: In function 'void __Pyx_AddTraceback(const char*, int, int, const char*)':
edenscmnative/traceprof.cpp:472:62: error: invalid use of incomplete type 'PyFrameObject' {aka 'struct _frame'}
  472 |   #define __Pyx_PyFrame_SetLineNumber(frame, lineno)  (frame)->f_lineno = (lineno)
      |                                                              ^~
edenscmnative/traceprof.cpp:4433:5: note: in expansion of macro '__Pyx_PyFrame_SetLineNumber'
 4433 |     __Pyx_PyFrame_SetLineNumber(py_frame, py_line);
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/python3.11/Python.h:42,
                 from edenscmnative/traceprof.cpp:32:
/usr/include/python3.11/pytypedefs.h:22:16: note: forward declaration of 'PyFrameObject' {aka 'struct _frame'}
   22 | typedef struct _frame PyFrameObject;
      |                ^~~~~~
edenscmnative/traceprof.cpp: In function 'void __Pyx_Coroutine_ExceptionClear(_PyErr_StackItem*)':
edenscmnative/traceprof.cpp:5586:20: error: '_PyErr_StackItem' {aka 'struct _err_stackitem'} has no member named 'exc_type'
 5586 |     t = exc_state->exc_type;
      |                    ^~~~~~~~
edenscmnative/traceprof.cpp:5588:21: error: '_PyErr_StackItem' {aka 'struct _err_stackitem'} has no member named 'exc_traceback'
 5588 |     tb = exc_state->exc_traceback;
      |                     ^~~~~~~~~~~~~
edenscmnative/traceprof.cpp:5589:16: error: '_PyErr_StackItem' {aka 'struct _err_stackitem'} has no member named 'exc_type'
 5589 |     exc_state->exc_type = NULL;
      |                ^~~~~~~~
edenscmnative/traceprof.cpp:5591:16: error: '_PyErr_StackItem' {aka 'struct _err_stackitem'} has no member named 'exc_traceback'
 5591 |     exc_state->exc_traceback = NULL;
      |                ^~~~~~~~~~~~~
edenscmnative/traceprof.cpp: In function 'PyObject* __Pyx_Coroutine_SendEx(__pyx_CoroutineObject*, PyObject*, int)':
edenscmnative/traceprof.cpp:5668:20: error: '_PyErr_StackItem' {aka 'struct _err_stackitem'} has no member named 'exc_type'
 5668 |     if (exc_state->exc_type) {
      |                    ^~~~~~~~
edenscmnative/traceprof.cpp:5671:24: error: '_PyErr_StackItem' {aka 'struct _err_stackitem'} has no member named 'exc_traceback'
 5671 |         if (exc_state->exc_traceback) {
      |                        ^~~~~~~~~~~~~
edenscmnative/traceprof.cpp:5672:70: error: '_PyErr_StackItem' {aka 'struct _err_stackitem'} has no member named 'exc_traceback'
 5672 |             PyTracebackObject *tb = (PyTracebackObject *) exc_state->exc_traceback;
      |                                                                      ^~~~~~~~~~~~~
edenscmnative/traceprof.cpp:5676:14: error: invalid use of incomplete type 'PyFrameObject' {aka 'struct _frame'}
 5676 |             f->f_back = PyThreadState_GetFrame(tstate);
      |              ^~
/usr/include/python3.11/pytypedefs.h:22:16: note: forward declaration of 'PyFrameObject' {aka 'struct _frame'}
   22 | typedef struct _frame PyFrameObject;
      |                ^~~~~~
edenscmnative/traceprof.cpp: In function 'void __Pyx_Coroutine_ResetFrameBackpointer(_PyErr_StackItem*)':
edenscmnative/traceprof.cpp:5707:35: error: '_PyErr_StackItem' {aka 'struct _err_stackitem'} has no member named 'exc_traceback'
 5707 |     PyObject *exc_tb = exc_state->exc_traceback;
      |                                   ^~~~~~~~~~~~~
In file included from /usr/include/python3.11/Python.h:38:
edenscmnative/traceprof.cpp:5713:19: error: invalid use of incomplete type 'PyFrameObject' {aka 'struct _frame'}
 5713 |         Py_CLEAR(f->f_back);
      |                   ^~
/usr/include/python3.11/pyport.h:24:38: note: in definition of macro '_Py_CAST'
   24 | #define _Py_CAST(type, expr) ((type)(expr))
      |                                      ^~~~
/usr/include/python3.11/object.h:581:29: note: in expansion of macro '_PyObject_CAST'
  581 |         PyObject *_py_tmp = _PyObject_CAST(op); \
      |                             ^~~~~~~~~~~~~~
edenscmnative/traceprof.cpp:5713:9: note: in expansion of macro 'Py_CLEAR'
 5713 |         Py_CLEAR(f->f_back);
      |         ^~~~~~~~
/usr/include/python3.11/pytypedefs.h:22:16: note: forward declaration of 'PyFrameObject' {aka 'struct _frame'}
   22 | typedef struct _frame PyFrameObject;
      |                ^~~~~~
In file included from /usr/include/python3.11/Python.h:44:
edenscmnative/traceprof.cpp:5713:19: error: invalid use of incomplete type 'PyFrameObject' {aka 'struct _frame'}
 5713 |         Py_CLEAR(f->f_back);
      |                   ^~
/usr/include/python3.11/object.h:583:14: note: in definition of macro 'Py_CLEAR'
  583 |             (op) = NULL;                        \
      |              ^~
/usr/include/python3.11/pytypedefs.h:22:16: note: forward declaration of 'PyFrameObject' {aka 'struct _frame'}
   22 | typedef struct _frame PyFrameObject;
      |                ^~~~~~
In file included from /usr/include/python3.11/Python.h:45:
edenscmnative/traceprof.cpp: In function 'int __Pyx_Coroutine_traverse_excstate(_PyErr_StackItem*, visitproc, void*)':
edenscmnative/traceprof.cpp:6019:25: error: '_PyErr_StackItem' {aka 'struct _err_stackitem'} has no member named 'exc_type'
 6019 |     Py_VISIT(exc_state->exc_type);
      |                         ^~~~~~~~
/usr/include/python3.11/objimpl.h:199:13: note: in definition of macro 'Py_VISIT'
  199 |         if (op) {                                                       \
      |             ^~
edenscmnative/traceprof.cpp:6019:25: error: '_PyErr_StackItem' {aka 'struct _err_stackitem'} has no member named 'exc_type'
 6019 |     Py_VISIT(exc_state->exc_type);
      |                         ^~~~~~~~
/usr/include/python3.11/pyport.h:24:38: note: in definition of macro '_Py_CAST'
   24 | #define _Py_CAST(type, expr) ((type)(expr))
      |                                      ^~~~
/usr/include/python3.11/objimpl.h:200:30: note: in expansion of macro '_PyObject_CAST'
  200 |             int vret = visit(_PyObject_CAST(op), arg);                  \
      |                              ^~~~~~~~~~~~~~
edenscmnative/traceprof.cpp:6019:5: note: in expansion of macro 'Py_VISIT'
 6019 |     Py_VISIT(exc_state->exc_type);
      |     ^~~~~~~~
edenscmnative/traceprof.cpp:6021:25: error: '_PyErr_StackItem' {aka 'struct _err_stackitem'} has no member named 'exc_traceback'
 6021 |     Py_VISIT(exc_state->exc_traceback);
      |                         ^~~~~~~~~~~~~
/usr/include/python3.11/objimpl.h:199:13: note: in definition of macro 'Py_VISIT'
  199 |         if (op) {                                                       \
      |             ^~
edenscmnative/traceprof.cpp:6021:25: error: '_PyErr_StackItem' {aka 'struct _err_stackitem'} has no member named 'exc_traceback'
 6021 |     Py_VISIT(exc_state->exc_traceback);
      |                         ^~~~~~~~~~~~~
/usr/include/python3.11/pyport.h:24:38: note: in definition of macro '_Py_CAST'
   24 | #define _Py_CAST(type, expr) ((type)(expr))
      |                                      ^~~~
/usr/include/python3.11/objimpl.h:200:30: note: in expansion of macro '_PyObject_CAST'
  200 |             int vret = visit(_PyObject_CAST(op), arg);                  \
      |                              ^~~~~~~~~~~~~~
edenscmnative/traceprof.cpp:6021:5: note: in expansion of macro 'Py_VISIT'
 6021 |     Py_VISIT(exc_state->exc_traceback);
      |     ^~~~~~~~
edenscmnative/traceprof.cpp: In function '__pyx_CoroutineObject* __Pyx__Coroutine_NewInit(__pyx_CoroutineObject*, __pyx_coroutine_body_t, PyObject*, PyObject*, PyObject*, PyObject*, PyObject*)':
edenscmnative/traceprof.cpp:6268:23: error: '_PyErr_StackItem' {aka 'struct _err_stackitem'} has no member named 'exc_type'
 6268 |     gen->gi_exc_state.exc_type = NULL;
      |                       ^~~~~~~~
edenscmnative/traceprof.cpp:6270:23: error: '_PyErr_StackItem' {aka 'struct _err_stackitem'} has no member named 'exc_traceback'
 6270 |     gen->gi_exc_state.exc_traceback = NULL;
      |                       ^~~~~~~~~~~~~
error: command '/usr/bin/gcc' failed with exit code 1
make: *** [Makefile:75: oss] Error 1
```

**Comment by leahneukirchen** [comment]
Worked for me after `git clean -fdx`, just `make clean` isn't enough.
